addons/binding/org.openhab.binding.lametrictime/src/main/java/org/openhab/binding/lametrictime/internal/discovery/LaMetricTimeAppDiscoveryService.java

```diff
@@ -0,0 +1,131 @@
+import org.eclipse.smarthome.config.discovery.DiscoveryResult;
+import org.eclipse.smarthome.config.discovery.DiscoveryResultBuilder;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.eclipse.smarthome.core.thing.binding.ThingHandler;
+import org.openhab.binding.lametrictime.LaMetricTimeBindingConstants;
+import org.openhab.binding.lametrictime.config.LaMetricTimeAppConfiguration;
+import org.openhab.binding.lametrictime.handler.LaMetricTimeAppHandler;
+import org.openhab.binding.lametrictime.handler.LaMetricTimeHandler;
+import org.openhab.binding.lametrictime.internal.LaMetricTimeUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.syphr.lametrictime.api.local.model.Application;
+import org.syphr.lametrictime.api.local.model.Widget;
+import org.syphr.lametrictime.api.model.CoreApps;
+
+import com.google.common.collect.Maps;
+import com.google.common.collect.Sets;
+
+/**
+ * The {@link LaMetricTimeAppDiscoveryService} is responsible for processing the
+ * list of apps found on the LaMetric Time device.
+ *
+ * @author Gregory Moyer - Initial contribution
+ */
+public class LaMetricTimeAppDiscoveryService extends AbstractDiscoveryService {
+
+    private static final Map<String, ThingTypeUID> CORE_APP_THING_TYPE_UIDS = Maps.newHashMap();
+    static {
+        CORE_APP_THING_TYPE_UIDS.put(CoreApps.clock().getPackageName(),
+                LaMetricTimeBindingConstants.THING_TYPE_CLOCK_APP);
+        CORE_APP_THING_TYPE_UIDS.put(CoreApps.countdown().getPackageName(),
+                LaMetricTimeBindingConstants.THING_TYPE_COUNTDOWN_APP);
+        CORE_APP_THING_TYPE_UIDS.put(CoreApps.radio().getPackageName(),
+                LaMetricTimeBindingConstants.THING_TYPE_RADIO_APP);
+        CORE_APP_THING_TYPE_UIDS.put(CoreApps.stopwatch().getPackageName(),
+                LaMetricTimeBindingConstants.THING_TYPE_STOPWATCH_APP);
+        CORE_APP_THING_TYPE_UIDS.put(CoreApps.weather().getPackageName(),
+                LaMetricTimeBindingConstants.THING_TYPE_WEATHER_APP);
+    }
+
+    private static final int TIMEOUT = 60;
+
+    private final Logger logger = LoggerFactory.getLogger(LaMetricTimeAppDiscoveryService.class);
+
+    private final LaMetricTimeHandler deviceHandler;
+
+    /**
+     * Discovers apps on the LaMetric Time device.
+     *
+     * @param deviceHandler the LaMetric Time device handler (bridge)
+     */
+    public LaMetricTimeAppDiscoveryService(final LaMetricTimeHandler deviceHandler) {
+        super(Sets.union(Sets.newHashSet(CORE_APP_THING_TYPE_UIDS.values()),
+                Collections.singleton(LaMetricTimeBindingConstants.THING_TYPE_GENERIC_APP)), TIMEOUT, true);
+        this.deviceHandler = deviceHandler;
+    }
+
+    @Override
+    protected void startBackgroundDiscovery() {
+        logger.debug("Starting background discovery for new apps");
+        startScan();
```

Comment:
A background discovery is supposed to run regularly and not just once. I am not sure though, whether we really need a regular scan as usually not much will change (and it isn't hard to retrigger a scan for the user, in case he has added a new app).
So I'd rather suggest to remove this method and pass a `false`in line 71.

One issue I noticed anyhow: The auto-discovery of the apps didn't work, I had to manually run a scan to make them appear. I assume this is because the call here happens before the handler has connected to the device.

It therefore might make sense to introduce an AppListener interface, with which the discovery service can register as a listener at the handler, so that the handler can pass all found apps to the listener once it is connected. Wdyt?

Reply:
It would probably be better to keep this simple and just remove the background scan letting the user hit the button to pick up new apps. The manual scan is working fine and like you said, it doesn't seem that hard for the user to hit the scan button when new apps are added that he/she would like to access in OH. An automatic scan for new apps can be an improvement later if there is demand, especially since we are removing generic app support entirely for now.